### DISCUSSION TITLE
Finding % of a specified color (convert)

### DISCUSSION THREAD

**Discussion by Nemesis77swe** [ImageMagick/ImageMagick#2056] [Help]
Hi all!

Brand New to ImageMagic.

I'm having issues with detecting percentage of a specified color (yes, even thou it says how on the ImageMagic document)

```
convert test.jpg -fill black +opaque blue \                    -fill white -opaque blue \                    -print "blue pixels = %[fx:w*h*mean]\n"   null:

```
Straight from the manual, except I've changed the color from yellow to blue, and a jpg instead of a gif.
This resulted in a black photo (0px blue)

Now, the blue in this picture:
![test](https://user-images.githubusercontent.com/40662061/82681654-60954980-9c4e-11ea-8e37-3a4b78e664e2.jpg)
Is very dark, so I did this:
```
convert test.jpg -negate \
-fill white +opaque blue \
                   -fill black -opaque blue \
                   -print "blue pixels = %[fx:w*h*mean]\n"  null:

```

And that made a white picture..

I thought I followed the description, but I must do something wrong..

**Comment by snibgo** [comment]
I suppose your input has no pixels that are exactly blue.

You can use "-fuzz 5%" or whatever.

**Comment by Nemesis77swe** [comment]
Oh, I tried to implement it, but I only got errors, where should I put it?

-- 
Mkkhcnll May 22, 2020 6:24:51 PM snibgo <notifications@github.com>:

**Comment by fmw42** [comment]
Post your original image. Does it have transparency? Did you use -fuzz XX% to adjust near colors? Your commands do not measure the percentage of color. They simply change colors. Do you really want to change colors or simply determine the percentage of some color in an image? JPG is a lossy compression and will change your colors. So I would suggest using GIF or PNG or TIFF for your output.

**Comment by Nemesis77swe** [comment]
I posted my original photo, it's from a screenshot..
No, I don't want to convert it, I only want to know how much blue there are in the photo (water), going to implant it in a bash scripts where it will scan all (40'000) pictures and copy those with 50%+ blue..

May 22, 2020 7:31:42 PM Fred Weinhaus <notifications@github.com>:

**Comment by fmw42** [comment]
I think what you might be trying to do is as follows for computing the percent of the medium blue or the count of medium blue pixels. Note that the medium blue is rgb(0,0,112), but is not perfectly constant, since the image is a JPG. So I add -fuzz 1% to allow for the slight variations.

`convert test.jpg -fuzz 1% -fill white -opaque "rgb(0,0,112)" -fuzz 0 -fill black +opaque white +write test_white.png -format "percent=%[fx:100*mean]\n" info:`

percent=5.87886

`convert test.jpg -fuzz 1% -fill white -opaque "rgb(0,0,112)" -fuzz 0 -fill black +opaque white +write test_white.png -format "count=%[fx:mean*w*h]\n" info:`

count=15231

**Comment by fmw42** [comment]
Your image does not have "blue" pixels that are rgb(0,0,255). The value is rgb(0,0,112). But is not perfectly constant since it is a JPG. So you need to add -fuzz XX% to allow for the slight variation and a closer value for the blue color in rgb(...).

`convert test.jpg -fuzz 1% -fill white -opaque "rgb(0,0,112)" -fuzz 0 -fill black +opaque white +write test_white.png -format "percent=%[fx:100*mean]\n" info:`

percent=5.87886

`convert test.jpg -fuzz 1% -fill white -opaque "rgb(0,0,112)" -fuzz 0 -fill black +opaque white +write test_white.png -format "count=%[fx:mean*w*h]\n" info:`

count=15231

In the above, I have also added +write to save the black and white image. So you can see where the blue was located.

![test_white](https://user-images.githubusercontent.com/13458218/82695683-6cffaf00-9c1a-11ea-8b4d-a2b94cd06fa9.png)

**Comment by fmw42** [comment]
Your image does not have perfectly "blue" rgb(0,0,255) colors. The medium blue is rgb(0,0,112). But it is not constant, since the image is a JPG and has lossy compression. So you need to use a closer blue color as rgb(...) and use a -fuzz XX% for tolerance.

`convert test.jpg -fuzz 7% -fill white -opaque "rgb(0,0,112)" -fuzz 0 -fill black +opaque white +write test_white.png -format "percent=%[fx:100*mean]\n" info:`

percent=14.8552

`convert test.jpg -fuzz 7% -fill white -opaque "rgb(0,0,112)" -fuzz 0 -fill black +opaque white +write test_white.png -format "count=%[fx:mean*w*h]\n" info:`

count=38487

I have added +write test_white.png to save the black and white image to show where the blue colors were identified.

![test_white](https://user-images.githubusercontent.com/13458218/82696097-38402780-9c1b-11ea-97cc-c819f037a1ca.png)

**Comment by Nemesis77swe** [comment]
Thx for the help, I actually tried:

convert test.jpg -fill black  -fuzz 20% +opaque navy \
-fill white  -opaque navy  \
-print "%[fx:w*h*mean]"   tata.jpg >txt.txt (based on your previous information)

And I actually think I got a lot of the blue (if you're comparing my attached screenshot with the original photo.
My issue is only if this would work on "all" images, I mean if I use 35% the image is getting whole white...

May 22, 2020 7:53:47 PM Fred Weinhaus <notifications@github.com>:

**Comment by fmw42** [comment]
> convert test.jpg -fill black  -fuzz 20% +opaque navy \
> -fill white  -opaque navy  \
> -print "%[fx:w*h*mean]"   tata.jpg >txt.txt (based on your previous information)
> 
> And I actually think I got a lot of the blue (if you're comparing my attached screenshot with the original photo.
> My issue is only if this would work on "all" images, I mean if I use 35% the image is getting whole white...
> 

No, you must tune to a near color and use an appropriate fuzz value so as not to miss any nor add too much of other colors.

**Comment by Nemesis77swe** [comment]
But isn't that what I just did?
I saw that the blue was navy instead of blue, and 20% fuzz of navy must stay pretty blue? (There are no black in the images)

May 22, 2020 8:24:05 PM Fred Weinhaus <notifications@github.com>:

**Comment by fmw42** [comment]
> But isn't that what I just did?
> I saw that the blue was navy instead of blue, and 20% fuzz of navy must stay pretty blue? (There are no black in the images)
> 

Yes. I was just pointing out that named colors may not be close enough forcing you to use too large a fuzz value. You should pick the colors as close as possible and use as small a fuzz value as you can. Alternately, you can try the -color-threshold method that lets you pick a range of colors.

**Comment by Nemesis77swe** [comment]
Ah, I c....
Well, thank you so much for the help!
I really appreciate it!

-- 
Mkkhcnll May 22, 2020 9:04:57 PM Fred Weinhaus <notifications@github.com>:

**Comment by snibgo** [comment]
See also the recent "Percentage of a color" thread: https://www.imagemagick.org/discourse-server/viewtopic.php?f=1&t=37876
